rfcs/NNNN-lit-3.0.md

```diff
@@ -0,0 +1,112 @@
+
+### Publishing ES2021
+
+We published Lit 1.0 packages as ES2019 because of spotty browser support for some ES2021 features like logical assignment. The situation is much better now, and publishing native optional chaining and logical assignment expressions will give us a small bundle size and perf improvement. 
+
+### Clean up deprecated APIs
+
+We need to start the habit of regularly removing previously deprecated APIs to keep the codebase clean.
+
+## Detailed Design
+
+### IE11 removal
+
+- Turn down our IE test runners
+- Remove workaround for IE normalizing text nodes during clone ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/lit-html/src/lit-html.ts#L974))
+- Simplify bound attribute handling ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/lit-html/src/lit-html.ts#L911))
+- Simplify `styleMap()` style removal ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/lit-html/src/directives/style-map.ts#L82))
+- Use `globalThis` ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/lit-html/src/lit-html.ts#L15))
+- Use `toggleAttribute()` ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/reactive-element/src/reactive-element.ts#L1126))
+- Use `Element.prototype.after()`, `.before()`, `.replaceChildren()` where beneficial
+
+### Publish ES2021
+
+- Update tsconfig target and lib to es2022
+- Remove unused catch bindings ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/reactive-element/src/reactive-element.ts#L351))
+
+### Misc / Old API removal
+- Remove `UpdatingElement` export from lit-element.ts ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/lit-element/src/lit-element.ts#L84))
+- Remove extra exports from lit-element/index.ts ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/lit-element/src/index.ts#L10))
+- Move SSR hydration support modules to `@lit-labs/ssr-client`
+- Remove `requestUpdateThenable` ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/reactive-element/src/reactive-element.ts#L42))
+- Remove old API warnings ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/reactive-element/src/reactive-element.ts#L810))
+
+
+## Implementation Considerations
+
+### Implementation Plan
+
+#### Branches
+
+If we keep the breaking changes to this version to only dropping IE11 support and publishing ES2020 we may be able to land the changes on `main` in quick succession with no special release branches.
+
+This may make importing into Google harder though - say if we batch the IE11 changes into one PR and that breaks internal Google tests. So we may want to make a `2.x` branch and commit PRs to `main` more incrementally.
```

Comment:
We need to look into how we want to do releases and how the changeset action behaves in these situations. I'm worried that if we don't come up with a strategy that handles branching that once we've started doing prereleases off of `main` we won't feel comfortable making regular releases until we've landed 3.0.

Reply:
I think we should definitely have separate branches for 2.x and 3.x. The main question is whether `main` should be 2.x for a short while.

I think the longer that work for 3.x takes (like if we wanted to start now, but not release for a few months) the more I'd lean towards `main` staying 2.x for a bit and making a branch for 3.x. I'd then want to make prototype PRs onto 3.x to test everything out and get most of the reviews out of the way, and at some point recreate the PRs onto `main` (which is hopefully cheap).

The other approach is to make `main` the 3.x / prerelease branch, and cherry-pick non-breaking PRs to the 2.x branch. I wouldn't want to be in that state for too long though.

So to me this kind of hinges on whether we want standard decorators in 3.0. That's the only thing we would have to wait for.

Reply:
We're following the changeset recommendation of a branch for the prereleases.